drafts/AENS.md

```diff
@@ -0,0 +1,414 @@
+The `.aet` namespace registrar allows users to claim any name that is
+valid according to the validation rules for labels and names if they
+are the first to claim the name and pay the appropriate fee.
+
+The `.test` namespace registrar can be, as the name suggests, used for
+test purposes
+
+
+### Commitment Scheme
+
+Claiming a name requires a commitment scheme, which enables the claimant
+to commit to a name and after the commitment has been accepted into the 
+chain, reveal the name to finish the process.
+
+A commitment should be binding, i.e. the claimant cannot change
+the value they commited to withouth changing the actual commitment
+and hiding, so that a malicious miner learns nothing about the value
+the claimant has commited until they chose to reveal that value. This
+prevents malicious miners from front running, i.e. upon seeing a 
+claim transaction, including their own claim request for the same
+name instead of the original claimant's one.
+
+
+### Hashing
+
+We are going to adopt the `NameHash` as defined by the ENS in 
+[EIP-137](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-137.md#namehash-algorithm)
+using SHA256 instead.
+
+```
+def namehash(name):
+  if name == '':
+    return '\0' * 32
+  else:
+    label, _, remainder = name.partition('.')
+    return sha256(namehash(remainder) + sha256(label))
+```
+
+Names are generally only referred to in hashed form.
+
+### Protocol
+
+```
+                  expire
+      unclaimed <-------- revoked
+          | ^              ^^
+          | |              || expire
+          | | expire       ||
+pre-claim | |              ||  _
+          | |       revoke || | | transfer
+          v |              || | v
+     pre-claimed -------> claimed
+                  claim    | ^   
+                           | |
+                            -
+                          update
+```
+
+Note that `expire` is not an explicit message that is part
+of the protocol.
+
+
+#### Pre-claim
+
+Send a pre-claim transaction containing a hash commitment.
+
+```
+ ------------ ----
+| hash       | 32 |
+ ------------ ----
+```
+
+
+
+#### Claim
+
+```
+ ------------ ----
+| name       | 63 |
+ ------------ ----
+| fee        | 32 |
+ ------------ ----
+```
+
+Flow for a user:
+
+1. (optional) wait `n` blocks, s.t. that the block including the `pre-claim` cannot be reversed whp 
+2. send `claim` transaction to reveal name and pay the associated fee
+
+If the time delta of 1 and 2 is bigger than e.g. 48 hours then the 
+`pre-claim` expires again.
+
+The `claim` transaction MUST be signed by the same private key as the 
+`pre-claim` transaction.
+
+
+#### Update
+
+```
+ ------------ ----
+| hash       | 32 |
+ ------------ ----
+| expire_by  | 8  |
+ ------------ ----
+| ttl        | 8  |
+ ------------ ----
+| pointers   |    |
+ ------------ ----
+```
+
+The `update` transaction MUST be signed by the owner
+of the name entry to be updated.
+
+The `expire_by` MUST NOT be more than 1 year into the
```

Comment:
To be consistent, maybe it is better to express this MUST NOT in terms of number of blocks ahead of current top block. Now we are mixing block height and wall clock time...

Reply:
Sorry, all times should be expressed in blocks. I should make that explicit and not just assume it.